### ISSUE TITLE
Differentiable views don't work well with python autograd functions

### ISSUE THREAD

**Issue by zou3519** [pytorch/pytorch#14707]
If the output of a python autograd function is a view and gets modified in-place with a torch.no_grad block later, the python autograd function's backward function doesn't get called and instead an autograd backward pass (through the forward function) runs. This results in incorrect behavior, especially when the forward pass contains non-differentiable things. 

Expected behavior: we should either throw an error message about something or compute the correct gradients. Neither of those happen right now.

```
import torch
from torch.autograd import Function

class Mul3ThenFlatten(Function):
    @staticmethod
    def forward(ctx, x):
        ctx.save_for_backward(x)
        z = x.detach().numpy() * 3
        # If I add a .contiguous() to the end of this statement, other things happen.
        return torch.tensor(z).view(-1)

    @staticmethod
    def backward(ctx, dx):
        x = ctx.saved_tensors[0]
        print("Some side effect")
        return dx.view(x.shape) * 3

x = torch.randn(2, 2, requires_grad=True)
output = Mul3ThenFlatten.apply(x)

# Modify the counter in no_grad mode.
with torch.no_grad():
    output.fill_(1)

# Doesn't print "Some side effect" :/
output.sum().backward()

x.grad  # is None, should really be something or at least an assert
```

cc @colesbury

**Comment by soumith** [comment]
@zou3519 @colesbury @ngimel @apaszke is this a release blocker? I need patches to land by tomorrow night if so

**Comment by albanD** [comment]
`.contiguous()` shouldn't do anything here as it is contiguous even after the view. Using `.clone()` should fix the issue though.

If you modify the `output` not in `no_grad` mode, you still have the same problem? Meaning that you make this change in a differentiable manner.

Isn't the problem simply that because of the way the view_as is done in the binding [here](https://github.com/pytorch/pytorch/blob/b15242f70cce10c0a0f8c36b690ac2481c51814f/tools/autograd/gen_variable_type.py#L540), all call to `.view()` will actually generate a differentiable view, even when grad_mode is disabled?
[This](https://github.com/pytorch/pytorch/blob/b15242f70cce10c0a0f8c36b690ac2481c51814f/tools/autograd/gen_variable_type.py#L560-L562) comment in the generation code is especially weird.
Changing the `true` by `bool(grad_fn)` in the generation code [here](https://github.com/pytorch/pytorch/blob/b15242f70cce10c0a0f8c36b690ac2481c51814f/tools/autograd/gen_variable_type.py#L562) and [here](https://github.com/pytorch/pytorch/blob/b15242f70cce10c0a0f8c36b690ac2481c51814f/tools/autograd/gen_variable_type.py#L542) does fix the issue presented here and proper gradients are computed. But I'm not sure if that's the proper fix here.

**Comment by colesbury** [comment]
```
If you modify the output not in no_grad mode, you still have the same problem? Meaning that you make this change in a differentiable manner.
```

If you modify the output in normal differentiable mode, this is reasonable behavior. The gradient of `output` w.r.t. `x` is zero.

**Comment by albanD** [comment]
@colesbury true for this case, but doing `output += 1` would break the graph the same way but the true gradients are not 0.

**Comment by zou3519** [comment]
If you modify the output in normal differentiable mode, the gradient of output w.r.t `x` is not zero @colesbury 
```
In [7]: import torch
   ...: from torch.autograd import Function
   ...:
   ...: class Mul3ThenFlatten(Function):
   ...:     @staticmethod
   ...:     def forward(ctx, x):
   ...:         ctx.save_for_backward(x)
   ...:         z = x.detach().numpy() * 3
   ...:         # If I add a .contiguous() to the end of this statement, other things happen.
   ...:         return torch.tensor(z).view(-1)
   ...:
   ...:     @staticmethod
   ...:     def backward(ctx, dx):
   ...:         x = ctx.saved_tensors[0]
   ...:         print("Some side effect")
   ...:         return dx.view(x.shape) * 3
   ...:
   ...: x = torch.randn(2, 2, requires_grad=True)
   ...: output = Mul3ThenFlatten.apply(x)
   ...:
   ...: # Doesn't print "Some side effect" :/
   ...: output.sum().backward()
   ...:
   ...: x.grad  # is None, should really be something or at least an assert
Some side effect
Out[7]:
tensor([[3., 3.],
        [3., 3.]])
```

**Comment by colesbury** [comment]
@zou3519 you removed the `output.fill_(1)` call

**Comment by zou3519** [comment]
Oh good point, please ignore my comment

**Comment by colesbury** [comment]
There's definitely at least one bug here.  I think one issue is that `python_function.cpp` is treating `torch.tensor(z).view(-1)` as a view of `torch.tensor(z)`, which isn't part of the autograd graph.

The second issue is that:

```
with torch.no_grad():
    output.fill_(1)
```

should probably not modify the autograd graph because it's in a no_grad() block.

**Comment by zou3519** [comment]
```
with torch.no_grad():
    output.fill_(1)
```
bumps the version counter on output. I don't think it modifies the graph otherwise.

**Comment by ezyang** [comment]
I was distinctly under the impression that https://github.com/pytorch/pytorch/pull/12502 fixes this. cc @SsnL

**Comment by SsnL** [comment]
@ezyang #12502 fixes views created in no_grad block. This is a different issue on the subtle interactions amongnpython autograd fn, inplace op on views, and our grad_fn replacement trick.

**Comment by zou3519** [comment]
I think if one implemented a native function like this that returned multiple views, it would have the same problem (I haven't tried doing this yet), so this might not be limited to python autograd functions.

**Comment by albanD** [comment]
Closing as duplicate of https://github.com/pytorch/pytorch/issues/26546 to reduce noise.
